contents/javascript/Javascript-modules.md

```diff
@@ -0,0 +1,220 @@
+
+CommonJS is mainly used in NodeJS applications, but you can also use it outside of NodeJS by using bundler. It does not need transpiler like ES6 modules. Nevertheless, if your project requirement does not allow you to use bundler, there is still a modules implementation that does not need any external tool.
+
+### Module Pattern
+
+This is the way JavaScript developers implement modules in the past before the inception of CommonJS and ES6 modules. Using a technique in JavaScript called [IIFE (Immediately Invoked Function Expression)](https://developer.mozilla.org/en-US/docs/Glossary/IIFE), they wrap their code in an IIFE.
+
+> An **IIFE (Immediately Invoked Function Expression)** is a JavaScript function that runs as soon as it is defined. <br>
+> Syntax: `(function() { statements })();`
+
+The example below will replicate the same examples above but using module pattern.
+
+To include the module in `index.js`, we use the `<script>` tag in HTML to import it.
+
+```html
+<!-- index.html -->
+
+<script src="./anExampleModule.js" />
+<script src="./index.js" />
+```
+
+```js
+// anExampleModule.js
+
+var anExampleModule = (function() {
+    var variableOne = 1;
+    var variableTwo = 2;
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    // We return an object with functions or variables
+    // that we want to be exposed
+    return {
+        sumOfVariable: sumOfVariable
+    }
+})()
+```
+
+```js
+// index.js
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+The result of importing the modules in the HTML will be as follow:
+```js
+// anExampleModule.js
+
+var anExampleModule = (function() {
+    var variableOne = 1;
+    var variableTwo = 2;
+
+    var sumOfVariable = function() {
+        return variableOne + variableTwo;
+    }
+
+    // We return an object with functions or variables
+    // that we want to be exposed
+    return {
+        sumOfVariable: sumOfVariable
+    }
+})()
+
+// index.js
+
+anExampleModule.sumOfVariable(); // 3
+```
+
+### Which to use?
+
+With the launch of ES6 modules, one might think that why should we choose which one to use? Why not use ES6 modules right away?
+
+As stated above, ES6 modules is currently not supported by all browsers without using transpiler and bundler. If an application does not allow you to use transpiler and bundler perhaps because of the additional overhead or you just need to create a simple and lean application, you might want to consider implementing module pattern. 
+
+ES6 modules is also currently not natively supported by NodeJS, which still uses CommonJS. This explains why a lot of NodeJS applications still use CommonJS.
```

Comment:
It feels like some things are mentioned 2-3 times in the article. This one is an example.